crates/forge_app/src/services.rs

```diff
@@ -112,6 +112,112 @@ pub trait FileDiscoveryService: Send + Sync {
     async fn collect(&self, max_depth: Option<usize>) -> anyhow::Result<Vec<File>>;
 }
 
+#[async_trait::async_trait]
+pub trait AttemptCompletion: Send + Sync {
+    /// Attempts to complete a tool call with the given context.
+    async fn attempt_completion(
+        &self,
+        result: String,
+        explanation: Option<String>,
+    ) -> anyhow::Result<String>;
+}
+
+#[async_trait::async_trait]
+pub trait FsCreate: Send + Sync {
+    /// Create a file at the specified path with the given content.
+    async fn create(
+        &self,
+        path: String,
+        content: String,
+        overwrite: bool,
+        explanation: Option<String>,
+    ) -> anyhow::Result<String>;
+}
+
+#[async_trait::async_trait]
+pub trait FsPatch: Send + Sync {
+    /// Patches a file at the specified path with the given content.
+    async fn patch(
+        &self,
+        path: String,
+        search: String,
+        operation: PatchOperation,
+        content: String,
+        explanation: Option<String>,
+    ) -> anyhow::Result<String>;
+}
+
+#[async_trait::async_trait]
+pub trait FsRead: Send + Sync {
+    /// Reads a file at the specified path and returns its content.
+    async fn read(
+        &self,
+        path: String,
+        start_line: Option<u64>,
+        end_line: Option<u64>,
```

Comment:
```suggestion
```

Reply:
Always read the full file via the service, and output format should handle trimming.